### PR TITLE
Adds generation of keys locally

### PR DESCRIPTION
**Pull request by nkakouros**
This is mainly a wrapper around `openssl_publickey` and `openssl_privatekey`.

### CODE REVIEW

**Comment by nkakouros** [comment]
The last two commits do not use the `openssl_privatekey` module but the `shell` module and calls `openssl` directly. I did that to allow the user to change the utility they use to generate the keys if they need to. This is sth that I use to set [keydgen](https://github.com/cornfeedhobo/ssh-keydgen) as the utility to generate deterministic keys.

**Comment by tterranigma** [comment]
Now that 2.8 is out, I think it would be a good idea to merge this. Since the deprecation cycle in ansible is 4 minor versions long, the oldest "officially" supported version is 2.5 now. The changes in this PR, I think should be compatible with at as old a version as 2.5.

File: tasks/known-hosts.yml [hunk 1]
```diff
@@ -16,9 +16,9 @@
     state: directory
     owner: root
     group: root
-    mode: 0755
```
Comment (tersmitten) [inline tasks/known-hosts.yml:19]: Why did you change this?

Comment (nkakouros) [reply tasks/known-hosts.yml:19]: ansible-lint (or yamllint, not sure) complains about hex numbers not in the `0oXXX` format.

**Comment by nkakouros** [comment]
@tersmitten Is there interest in merging this?

File: vars/main.yml [hunk 2]
```diff
@@ -4,3 +4,11 @@ ssh_keys_sshdir: '.ssh'
 ssh_keys_private_key_filename: 'id_rsa'
 ssh_keys_public_key_filename: 'id_rsa.pub'
 ssh_keys_known_hosts_path: /etc/ssh/ssh_known_hosts
+_ssh_keys_generate_command: >-
+  openssl \
+    gen{{ item.type | default('rsa') }} \
+    -out {{ item.path }} \
+    {{ '-' + item.cipher if 'cipher' in item else '' }} \
+    {{ '-passout stdin' if 'passphrase' in item else '' }} \
+    {{ item.size | default(2048) }} \
```
Comment (tersmitten) [inline vars/main.yml]: `4096`?

File: vars/main.yml [hunk 3]
```diff
@@ -4,3 +4,11 @@ ssh_keys_sshdir: '.ssh'
...
+    {{ '<<< "' + item.passphrase + '"' if 'passphrase' in item else '' }}
```
Comment (tersmitten) [inline vars/main.yml]: Can you rewrite this to use `ternary`?

File: README.md [hunk 4]
```diff
@@ -11,9 +11,25 @@ None
 
 #### Variables
 
+* `ssh_keys_generate_keys`: [default: `[]`]: Keys to generate locally
+* `ssh_keys_generate_keys.{n}.path`: [required] The local path where the key should be generated
+* `ssh_keys_generate_keys.{n}.cipher`: [optional, default: None]: The cipher to
+  encrypt the private key
+* `ssh_keys_generate_keys.{n}.format`: [optional, default: `PEM`]: The format of
+  the public key
+* `ssh_keys_generate_keys.{n}.force`: [optional, default: no]: Whether to
```
Comment (tersmitten) [inline README.md]: `false`

File: README.md [hunk 5]
```diff
@@ -11,9 +11,25 @@ None
...
+* `ssh_keys_generate_keys.{n}.mode`: [optional, default `0600`]: The UNIX permission mode bits of the key pair
+* `ssh_keys_generate_keys.{n}.passphrase`: [optional, default `''`]: The passphrase for the private key
+* `ssh_keys_generate_keys.{n}.size`: [optional, default `4096`]: Size in bits of the TLS/SSL key to generate
+* `ssh_keys_generate_keys.{n}.type`: [optional, default `RSA`]: The algorithm used to generate the private key
+
 * `ssh_keys_private_keys`: [default: `[]`]: Private key declarations
 * `ssh_keys_private_keys.{n}.owner`: [required]: The name of the user that should own the file
-* `ssh_keys_private_keys.{n}.group`: [optional, default `owner`]: The name of the group that should own the file
+* `ssh_keys_private_keys.{n}.group`: [optional, default `owner`]: The name of
+  the group that should own the file
```
Comment (tersmitten) [inline README.md]: Remove `newline`?

File: tasks/generate.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        force: "{{ item.force | default(false) }}"
+        format: "{{ item.format | default('OpenSSH') }}"
+        # comment: "{{ item.comment | default(omit) }}"
+        owner: "{{ item.owner }}"
+        group: "{{ item.group | default(item.owner) }}"
+        mode: "{{ item.mode | default(0600) }}"
+        privatekey_passphrase: "{{ item.passphrase | default(omit) }}"
+        privatekey_path: "{{ item.path }}"
+      with_items: "{{ ssh_keys_generate_keys }}"
+  delegate_to: localhost
```
Comment (tersmitten) [inline tasks/generate.yml]: I prefer `connection: local`

File: tasks/generate.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        item.force | default(false) | bool or
+        item.path is not file
+      with_items: "{{ ssh_keys_generate_keys }}"
+
+    - name: Set permissions on keys
+      file:
+        path: "{{ item.path }}"
+        owner: "{{ item.owner }}"
+        group: "{{ item.group | default(item.owner) }}"
+        mode: "{{ item.mode | default('0600') }}"
```
Comment (tersmitten) [inline tasks/generate.yml:27]: `0o600`?

File: tasks/generate.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+    - name: Generate public ssh key
+      openssl_publickey:
+        path: "{{ item.path }}.pub"
+        force: "{{ item.force | default(false) }}"
+        format: "{{ item.format | default('OpenSSH') }}"
+        # comment: "{{ item.comment | default(omit) }}"
+        owner: "{{ item.owner }}"
+        group: "{{ item.group | default(item.owner) }}"
+        mode: "{{ item.mode | default(0600) }}"
```
Comment (tersmitten) [inline tasks/generate.yml:38]: `0o600`?

File: README.md [hunk 9]
```diff
@@ -11,9 +11,25 @@ None
 
 #### Variables
 
+* `ssh_keys_generate_keys`: [default: `[]`]: Keys to generate locally
+* `ssh_keys_generate_keys.{n}.path`: [required] The local path where the key should be generated
+* `ssh_keys_generate_keys.{n}.cipher`: [optional, default: None]: The cipher to
+  encrypt the private key
+* `ssh_keys_generate_keys.{n}.format`: [optional, default: `PEM`]: The format of
```
Comment (tersmitten) [inline README.md]: `OpenSSH`?

File: tasks/generate.yml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        group: "{{ item.group | default(item.owner) }}"
+        mode: "{{ item.mode | default('0600') }}"
+      with_items: "{{ ssh_keys_generate_keys }}"
+
+    - name: Generate public ssh key
+      openssl_publickey:
+        path: "{{ item.path }}.pub"
+        force: "{{ item.force | default(false) }}"
+        format: "{{ item.format | default('OpenSSH') }}"
+        # comment: "{{ item.comment | default(omit) }}"
```
Comment (tersmitten) [inline tasks/generate.yml]: No `comment`?

**Comment by tersmitten** [comment]
Yes, absolutely. It looks awesome!

File: tasks/generate.yml (hunk 6 above)
Comment (tersmitten) [reply tasks/generate.yml]: And can you add `tags` and make the task `name`s consistent with the other files

File: tasks/generate.yml (hunk 8 above)
Comment (tersmitten) [inline tasks/generate.yml]: `mode: "{{ item.mode | default('0600') }}"`

File: tasks/known-hosts.yml (hunk 1 above)
Comment (tersmitten) [reply tasks/known-hosts.yml:19]: I prefer to keep them as they were. I cannot reproduce the linter issues

**Comment by nkakouros** [comment]
I fixed most issues.

File: tasks/generate.yml (hunk 6 above)
Comment (nkakouros) [reply tasks/generate.yml]: Done.

File: README.md (hunk 4 above)
Comment (nkakouros) [reply README.md]: Fixed

File: vars/main.yml (hunk 3 above)
Comment (nkakouros) [reply vars/main.yml]: Fixed

File: vars/main.yml (hunk 2 above)
Comment (nkakouros) [reply vars/main.yml]: Fixed

File: tasks/generate.yml (hunk 10 above)
Comment (nkakouros) [reply tasks/generate.yml]: I have forked the `openssl_publickey` module to add a `comment` key. I removed the commented line.
